diff --git a/pkg/dds/lib/src/dap/adapters/dart.dart b/pkg/dds/lib/src/dap/adapters/dart.dart
index 2886f56ca95..3b135b9622c 100644
--- a/pkg/dds/lib/src/dap/adapters/dart.dart
+++ b/pkg/dds/lib/src/dap/adapters/dart.dart
@@ -2052,11 +2052,13 @@ abstract class DartDebugAdapter<TL extends LaunchRequestArguments,
       if (vars != null) {
         Future<Variable> convert(int index, vm.BoundVariable variable) {
           var name = variable.name;
+          final declaredType = variable.declaredType;
+          final isNullable = _converter.isDeclaredNullable(declaredType);
           // Store the expression that gets this object as we may need it to
           // compute evaluateNames for child objects later.
           final value = variable.value;
           if (value is vm.InstanceRef) {
-            storeEvaluateName(value, name);
+            storeEvaluateName(value, name, isNullable: isNullable);
           }
           return _converter.convertVmResponseToVariable(
             thread,
@@ -2065,7 +2067,8 @@ abstract class DartDebugAdapter<TL extends LaunchRequestArguments,
             allowCallingToString: evaluateToStringInDebugViews &&
                 index < maxToStringsPerEvaluation,
             evaluateName:
-                name != null ? (evaluateName: name, isNullable: false) : null,
+                name != null ? (evaluateName: name, isNullable: isNullable) : null,
+            declaredTypeName: declaredType?.name,
             format: format,
           );
         }
diff --git a/pkg/dds/lib/src/dap/protocol_converter.dart b/pkg/dds/lib/src/dap/protocol_converter.dart
index 8bb92f43b38..a7c24a5abbb 100644
--- a/pkg/dds/lib/src/dap/protocol_converter.dart
+++ b/pkg/dds/lib/src/dap/protocol_converter.dart
@@ -488,6 +488,7 @@ class ProtocolConverter {
     required String? name,
     required EvaluateName? evaluateName,
     required bool allowCallingToString,
+    String? declaredTypeName,
     bool allowTruncatedValue = true,
     VariableFormat? format,
   }) async {
@@ -501,6 +502,7 @@ class ProtocolConverter {
       return dap.Variable(
         name: name ?? response.kind.toString(),
         evaluateName: evaluateName?.evaluateName,
+        type: declaredTypeName,
         value: await convertVmResponseToDisplayString(
           thread,
           response,
diff --git a/pkg/vm_service/lib/src/vm_service.dart b/pkg/vm_service/lib/src/vm_service.dart
index 10677dc3154..1e1894a5083 100644
--- a/pkg/vm_service/lib/src/vm_service.dart
+++ b/pkg/vm_service/lib/src/vm_service.dart
@@ -27,7 +27,7 @@ export 'snapshot_graph.dart'
         HeapSnapshotObjectNoData,
         HeapSnapshotObjectNullData;
 
-const String vmServiceVersion = '4.20.0';
+const String vmServiceVersion = '4.21.0';
 
 /// @optional
 const String optional = 'optional';
@@ -2769,6 +2769,9 @@ class BoundVariable extends Response {
   /// [value] can be one of [InstanceRef], [TypeArgumentsRef] or [Sentinel].
   dynamic value;
 
+  /// The declared type of this variable.
+  InstanceRef? declaredType;
+
   /// The token position where this variable was declared.
   int? declarationTokenPos;
 
@@ -2781,6 +2784,7 @@ class BoundVariable extends Response {
   BoundVariable({
     this.name,
     this.value,
+    this.declaredType,
     this.declarationTokenPos,
     this.scopeStartTokenPos,
     this.scopeEndTokenPos,
@@ -2790,6 +2794,9 @@ class BoundVariable extends Response {
       : name = json['name'] ?? '',
         value = createServiceObject(json['value'],
             const ['InstanceRef', 'TypeArgumentsRef', 'Sentinel']),
+        declaredType =
+            createServiceObject(json['declaredType'], const ['InstanceRef'])
+                as InstanceRef?,
         declarationTokenPos = json['declarationTokenPos'] ?? -1,
         scopeStartTokenPos = json['scopeStartTokenPos'] ?? -1,
         scopeEndTokenPos = json['scopeEndTokenPos'] ?? -1,
@@ -2803,6 +2810,7 @@ class BoundVariable extends Response {
         'type': type,
         'name': name ?? '',
         'value': value?.toJson(),
+        'declaredType': declaredType?.toJson(),
         'declarationTokenPos': declarationTokenPos ?? -1,
         'scopeStartTokenPos': scopeStartTokenPos ?? -1,
         'scopeEndTokenPos': scopeEndTokenPos ?? -1,
@@ -2810,7 +2818,8 @@ class BoundVariable extends Response {
 
   @override
   String toString() => '[BoundVariable ' //
-      'name: $name, value: $value, declarationTokenPos: $declarationTokenPos, ' //
+      'name: $name, value: $value, declaredType: $declaredType, ' //
+      'declarationTokenPos: $declarationTokenPos, ' //
       'scopeStartTokenPos: $scopeStartTokenPos, scopeEndTokenPos: $scopeEndTokenPos]';
 }
 
