diff --git a/pkg/dds/lib/src/dap/adapters/dart.dart b/pkg/dds/lib/src/dap/adapters/dart.dart
index 2886f56ca95..3b135b9622c 100644
--- a/pkg/dds/lib/src/dap/adapters/dart.dart
+++ b/pkg/dds/lib/src/dap/adapters/dart.dart
@@ -2052,11 +2052,13 @@ abstract class DartDebugAdapter<TL extends LaunchRequestArguments,
       if (vars != null) {
         Future<Variable> convert(int index, vm.BoundVariable variable) {
           var name = variable.name;
+          final declaredType = variable.declaredType;
+          final isNullable = _converter.isDeclaredNullable(declaredType);
           // Store the expression that gets this object as we may need it to
           // compute evaluateNames for child objects later.
           final value = variable.value;
           if (value is vm.InstanceRef) {
-            storeEvaluateName(value, name);
+            storeEvaluateName(value, name, isNullable: isNullable);
           }
           return _converter.convertVmResponseToVariable(
             thread,
@@ -2065,7 +2067,8 @@ abstract class DartDebugAdapter<TL extends LaunchRequestArguments,
             allowCallingToString: evaluateToStringInDebugViews &&
                 index < maxToStringsPerEvaluation,
             evaluateName:
-                name != null ? (evaluateName: name, isNullable: false) : null,
+                name != null ? (evaluateName: name, isNullable: isNullable) : null,
+            declaredTypeName: declaredType?.name,
             format: format,
           );
         }
diff --git a/pkg/dds/lib/src/dap/protocol_converter.dart b/pkg/dds/lib/src/dap/protocol_converter.dart
index 8bb92f43b38..a7c24a5abbb 100644
--- a/pkg/dds/lib/src/dap/protocol_converter.dart
+++ b/pkg/dds/lib/src/dap/protocol_converter.dart
@@ -488,6 +488,7 @@ class ProtocolConverter {
     required String? name,
     required EvaluateName? evaluateName,
     required bool allowCallingToString,
+    String? declaredTypeName,
     bool allowTruncatedValue = true,
     VariableFormat? format,
   }) async {
@@ -501,6 +502,7 @@ class ProtocolConverter {
       return dap.Variable(
         name: name ?? response.kind.toString(),
         evaluateName: evaluateName?.evaluateName,
+        type: declaredTypeName,
         value: await convertVmResponseToDisplayString(
           thread,
           response,
